stream-chat-android-core/src/main/java/io/getstream/chat/android/client/call/MapCall.kt

```diff
@@ -16,45 +16,41 @@
 
 package io.getstream.chat.android.client.call
 
+import io.getstream.chat.android.client.call.Call.Companion.callCanceledError
 import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.client.utils.map
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withContext
 import java.util.concurrent.atomic.AtomicBoolean
 
 internal class MapCall<T : Any, K : Any>(
     private val call: Call<T>,
     private val mapper: (T) -> K,
 ) : Call<K> {
 
-    private var canceled = AtomicBoolean(false)
+    private val canceled = AtomicBoolean(false)
 
     override fun cancel() {
         canceled.set(true)
+        call.cancel()
     }
 
-    override fun execute(): Result<K> {
-        val resultA = call.execute()
+    override fun execute(): Result<K> = runBlocking { await() }
 
-        return if (resultA.isSuccess) {
-            val data = mapper(resultA.data())
-            Result(data)
-        } else {
-            val error = resultA.error()
-            Result(error)
+    override fun enqueue(callback: Call.Callback<K>) {
+        call.enqueue {
+            it.takeUnless { canceled.get() }
+                ?.map(mapper)
+                ?.let(callback::onResult)
         }
     }
 
-    override fun enqueue(callback: Call.Callback<K>) {
-        call.enqueue callback@{
-            if (canceled.get()) {
-                return@callback
-            }
-
-            if (it.isSuccess) {
-                val data = mapper(it.data())
-                callback.onResult(Result(data))
-            } else {
-                val error = it.error()
-                callback.onResult(Result(error))
-            }
-        }
+    override suspend fun await(): Result<K> = withContext(DispatcherProvider.IO) {
+        call.await()
```

Comment:
Just out of my curiosity, is there any reason that we don't use the injected coroutine scope for `MapCall` and `ZipCall`?

Reply:
We are not injecting the scope because it will complicate how the extensions to "map" or "zip" calls are used.
Currently we do something like that:
```
val callA: Call<Int> = [...] 
val callB: Call<String> = callA.map { "$it" }
```
Maybe from this part is not easy to access to and inject the proper coroutine scope